src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,451 @@
+        plural = isinstance(field_types, list)
+        if plural:
+            field_types = one(field_types)
+        if isinstance(field_types, dict):
+            yield {
+                "name": entity_type,
+                "type": {
+                    # This is always an array, even if singleton is passed in
+                    "type": "array",
+                    "items": {
+                        "name": entity_type,
+                        "type": "record",
+                        "fields": list(_tables_schema_recursive(field_types, files_entity=files_entity))
+                    }
+                }
+            }
+        elif field_types in nullable_to_pfb_types:
+            # Exceptions are fields that do not become lists during aggregation
+            exceptions = ('donor_count', 'total_estimated_cells')
+            if files_entity and not plural or entity_type in exceptions:
+                yield {
+                    "name": entity_type,
+                    "type": list(nullable_to_pfb_types[field_types]),
+                }
+            else:
+                yield {
+                    "name": entity_type,
+                    "type": {
+                        "type": "array",
+                        "items": list(nullable_to_pfb_types[field_types]),
+                    }
+                }
+        elif field_types is pass_thru_uuid4:
+            yield {
+                "name": entity_type,
+                "default": None,
+                # FIXME: Why does pass_thru_uuid appear in documents when other pass_throughs don't?
+                "type": ["string"],
+                "logicalType": "UUID"
+            }
+        elif field_types is value_and_unit:
+            yield {
+                "name": entity_type,
+                "type": {
+                    "name": entity_type,
+                    "type": "array",
+                    "items": [
+                        # Change 'string' to 'null' in https://github.com/DataBiosphere/azul/issues/2462
+                        "string",
+                        {
+                            "name": entity_type,
+                            "type": "record",
+                            "fields": [
+                                {
+                                    "name": name,
+                                    # Although, not technically a null_str, it's effectively the same
+                                    "type": nullable_to_pfb_types[null_str]
+                                } for name in ('value', 'unit')
+                            ]
+                        }
+                    ]
+                }
+            }
+        elif field_types in (pass_thru_json, pass_thru_int):
+            # Pass thru types are used only for aggregation and are excluded
+            # from actual hits
+            continue
+        else:
+            raise ValueError(f'Unexpected type in schema: {field_types!r}')
```

Comment:
```suggestion
            raise ValueError(f'Unexpected field type', field_types)
```

Wouldn't this constitute a bug? If so it should be an assertion.